### PR TITLE
Add support for specifying Snakebite effective user

### PR DESCRIPTION
**Pull request by RobCherry**
Adding a configuration option for specifying the effective_user parameter for the Snakebite clients.

### CODE REVIEW

File: luigi/hdfs.py [hunk 1]
```diff
@@ -285,27 +285,21 @@ def get_bite(self):
         If Luigi has forked, we have a different PID, and need to reconnect.
         """
         if self.pid != os.getpid() or not self._bite:
-            autoconfig_enabled = self.config.getboolean("hdfs", "snakebite_autoconfig", False)
-            if autoconfig_enabled is True:
+            client_kwargs = dict(filter(lambda (k, v): v is not None and v != '',  {
+                'hadoop_version': self.config.getint("hdfs", "client_version", None),
+                'effective_user': self.config.get("hdfs", "effective_user", None)
+            }.items()))
```
Comment (Tarrasch) [inline luigi/hdfs.py]: Thanks for respecting 2.6 backwards compatibility (by not using dict comprehensions) :)

File: luigi/hdfs.py [hunk 2]
```diff
@@ -285,27 +285,21 @@ def get_bite(self):
...
-                    ver = self.config.getint("hdfs", "client_version")
-                    if ver is None:
-                        raise RuntimeError()
-                    self._bite = Client(self.config.get("hdfs", "namenode_host"),
-                                       self.config.getint("hdfs", "namenode_port"),
-                                       hadoop_version=ver)
-                except:
-                    self._bite = Client(self.config.get("hdfs", "namenode_host"),
-                                       self.config.getint("hdfs", "namenode_port"))
+                self._bite = Client(self.config.get("hdfs", "namenode_host"), self.config.getint("hdfs", "namenode_port"), **client_kwargs)
```
Comment (Tarrasch) [inline luigi/hdfs.py]: Quite some changes here. Have you tested that the new stuff works?

File: luigi/hdfs.py [hunk 3]
```diff
@@ -285,27 +285,21 @@ def get_bite(self):
...
+            }.items()))
+            if self.config.getboolean("hdfs", "snakebite_autoconfig", False):
                 """
                 This is fully backwards compatible with the vanilla Client and can be used for a non HA cluster as well.
                 This client tries to read ``${HADOOP_PATH}/conf/hdfs-site.xml`` to get the address of the namenode.
                 The behaviour is the same as Client.
                 """
                 from snakebite.client import AutoConfigClient
-                self._bite = AutoConfigClient()
+                self._bite = AutoConfigClient(**client_kwargs)
```
Comment (Tarrasch) [inline luigi/hdfs.py]: Not related to this patch, but do you see why it uses the `AutoConfigClient` when `snakebite_autoconfig==False`?

Comment (RobCherry) [inline luigi/hdfs.py]: I don't follow.

File: luigi/hdfs.py (hunk 2 above)
Comment (RobCherry) [inline luigi/hdfs.py]: Unless there was some other exception that is intended to be caught and result in a new Client without the specified version, yes.

Comment (Tarrasch) [inline luigi/hdfs.py]: Yea, that code was quite badly written. Nice refactoring! :)

File: luigi/hdfs.py (hunk 3 above)
Comment (Tarrasch) [inline luigi/hdfs.py]: Eh... I think github is buggy or something. Why is the code **different** in the conversation view [1] and the diff view????

[1]: https://github.com/spotify/luigi/pull/594#discussion_r22802079
[2]: https://github.com/spotify/luigi/pull/594/files#r22802079

Anyway. Your code change makes total sense, I just misread the `False` in the if guard to be ` == False`, but of course that's not what it says. :)

**Comment by erikbern** [comment]
LGTM

**Comment by Tarrasch** [comment]
I'll make changes to the snakebite code myself today. So this must merge now (and everyone seems happy with it). :)

**Comment by Tarrasch** [comment]
Thank you very much @RobCherry !

**Pull request merged**